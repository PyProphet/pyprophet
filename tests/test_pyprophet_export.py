from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout


def _run_osw(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # peptide-level
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=global"

    # protein-level
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=global"

    # export
    cmdline += " export --in=test_data.osw --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    if not peptide:
        cmdline += " --no-peptide"

    if not protein:
        cmdline += " --no-protein"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_osw_unscored(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # export
    cmdline = "pyprophet export --in=test_data.osw --out=test_data.tsv --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_ipf(regtest, temp_folder, transition_quantification=False, ipf="disable"):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # MS2-level
    cmdline += " score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # IPF
    cmdline += " ipf --in=test_data.osw"

    # export
    cmdline += " export --in=test_data.osw --no-peptide --no-protein --ipf_max_peptidoform_pep=1  --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    cmdline += " --ipf=" + ipf

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_unscored(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # export
    cmdline = "pyprophet export-compound --in=test_data_compound.osw --out=test_data_compound_unscored.tsv --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_unscored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_compound_ms1(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms1.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms1.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_ms2(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS2-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms2 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms2.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms2.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)


def test_osw_0(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, False)

def test_osw_1(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, True, False, False)

def test_osw_2(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, True, False)

def test_osw_3(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, True)

def test_ipf_0(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "disable")

def test_ipf_1(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, True, "disable")

def test_ipf_2(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "peptidoform")

def test_ipf_3(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "augmented")

def test_compound_0(tmpdir, regtest):
    _run_compound_ms1(regtest, tmpdir.strpath)
    
def test_compound_1(tmpdir, regtest):
    _run_compound_ms2(regtest, tmpdir.strpath)

def test_compound_unscored(tmpdir, regtest):
	_run_compound_unscored(regtest, tmpdir.strpath)
	
def test_osw_unscored(tmpdir, regtest):
	_run_osw_unscored(regtest, tmpdir.strpath)
