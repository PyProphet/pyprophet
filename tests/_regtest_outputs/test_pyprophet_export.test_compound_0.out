   Adducts  Charge   Intensity       RT  aggr_prec_Peak_Apex  aggr_prec_Peak_Area  assay_RT  assay_rt     compound_name  d_score  decoy  delta_RT  delta_rt                                           filename                   id  leftWidth  m_score        mz  peak_group_rank  rightWidth               run_id    sum_formula  transition_group_id
0     M+H+     NaN  1.6598e+06  364.443           1.6824e+06           7.8688e+06   364.752  364.7520              Deet   3.7782      0   -0.3090   -0.3090  /Users/alka/Documents/work/projects/OpenSWATH/...  7823124520738135742    359.277   0.0313  192.1383                1     370.976   618490598098719992       C12H17NO                    0
1     M+H+     NaN  1.2839e+06  373.516           3.9995e+04           1.9509e+05   373.952  373.9524      Trimethacarb   4.2613      0   -0.4364   -0.4364  /Users/alka/Documents/work/projects/OpenSWATH/...  3484488607134072058    369.171   0.0149  194.1176                1     379.970  1532805053199387756      C11H15NO2                    8
2     M+H+     NaN  5.9785e+05  252.459           4.1461e+05           1.7107e+06   252.795  252.7951       Chloridazon   4.3396      0   -0.3361   -0.3361  /Users/alka/Documents/work/projects/OpenSWATH/...  6634143567532174618    247.782   0.0149  222.0429                1     260.382   618490598098719992     C10H8ClN3O                   12
3     M+H+     NaN  1.3043e+06  303.599           1.4001e+06           5.4964e+06   303.748  303.7483          Imazalil   4.4177      0   -0.1493   -0.1493  /Users/alka/Documents/work/projects/OpenSWATH/...  4509646909945349689    299.273   0.0086  297.0556                1     309.173  1277092869765949483   C14H14Cl2N2O                   16
4     M+H+     NaN  5.6452e+06  229.200           1.4137e+06           5.2582e+06   230.038  230.0376        Pirimicarb   4.2826      0   -0.8376   -0.8376  /Users/alka/Documents/work/projects/OpenSWATH/...  6687517813121524240    225.280   0.0149  239.1503                1     236.080  5672254479900394037     C11H18N4O2                   20
..     ...     ...         ...      ...                  ...                  ...       ...       ...               ...      ...    ...       ...       ...                                                ...                  ...        ...      ...       ...              ...         ...                  ...            ...                  ...
95    M+H+     NaN  4.8343e+06  484.464           1.2169e+06           5.7879e+06   484.380  484.3798  Pirmiphos-methyl   4.8472      0    0.0842    0.0842  /Users/alka/Documents/work/projects/OpenSWATH/...  7474895396820623250    480.156   0.0086  306.1036                1     490.056  5672254479900394037   C11H20N3O3PS                  306
96   M+Na+     NaN  1.1553e+07  538.227           1.6929e+05           9.3324e+05   538.751  538.7509      Furathiocarb   3.6450      0   -0.5239   -0.5239  /Users/alka/Documents/work/projects/OpenSWATH/...  6258159614040527067    533.542   0.0405  405.1455                1     543.441  1532805053199387756    C18H26N2O5S                  312
97   M+Na+     NaN  1.0354e+06  468.139           1.2596e+05           6.1140e+05   468.241  468.2407      Tebufenozide   4.1447      0   -0.1017   -0.1017  /Users/alka/Documents/work/projects/OpenSWATH/...  5376211212444909294    463.350   0.0220  375.2043                1     475.949   256173130911795889     C22H28N2O2                  318
98    M+H+     NaN  2.8745e+06  496.395           3.2453e+05           1.1530e+06   497.174  497.1739         Coumaphos   5.0961      0   -0.7789   -0.7789  /Users/alka/Documents/work/projects/OpenSWATH/...  6590669578008623220    492.054   0.0086  363.0217                1     501.053  1277092869765949483   C14H16ClO5PS                  322
99    M+H+     NaN  6.0873e+06  455.742           3.7840e+05           1.1234e+06   456.872  456.8716        Flufenacet   4.4102      0   -1.1296   -1.1296  /Users/alka/Documents/work/projects/OpenSWATH/...  7569401740039158287    451.550   0.0086  364.0737                1     460.550  5002162903887254472  C14H13F4N3O2S                  324

[100 rows x 23 columns]
